File: app/src/main/java/com/sysu/edu/academic/GradeForLevelActivity.java
package com.sysu.edu.academic;

import static com.sysu.edu.api.CommonUtil.extractValue;
import static com.sysu.edu.api.CommonUtil.isEmpty;
import static com.sysu.edu.api.CommonUtil.toStringOrDefault;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.widget.PopupWindow;

import androidx.appcompat.widget.PopupMenu;
import androidx.core.app.ActivityOptionsCompat;
import androidx.lifecycle.MutableLiveData;

import com.alibaba.fastjson2.JSONObject;
import com.google.android.material.button.MaterialButton;
import com.sysu.edu.BaseActivity;
import com.sysu.edu.R;
import com.sysu.edu.api.Params;
import com.sysu.edu.databinding.ActivityGradeForLevelBinding;
import com.sysu.edu.databinding.PreferenceEditBinding;
import com.sysu.edu.model.JwxtModel;
import com.sysu.edu.view.StaggeredFragment;

import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

public class GradeForLevelActivity extends BaseActivity {
    
    final MutableLiveData<String> trainType = new MutableLiveData<>();
    final MutableLiveData<String> year = new MutableLiveData<>();
    final MutableLiveData<String> courseType = new MutableLiveData<>();
    final MutableLiveData<String> courseName = new MutableLiveData<>();
    final MutableLiveData<String> courseNumber = new MutableLiveData<>();
    final MutableLiveData<String> minGrade = new MutableLiveData<>();
    //    HttpManager http;
    int page = 1;
    int total = -1;
    StaggeredFragment fragment;
    PopupMenu yearPop;
    PopupMenu trainTypePop;
    PopupMenu courseTypePop;
    MutableLiveData<String> input;
    JwxtModel model;
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        model.dispose();
    }
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityGradeForLevelBinding binding = ActivityGradeForLevelBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        model = new JwxtModel(this);
        Params params = new Params(this);
        params.setCallback(() -> {
            getData(0);
            regetGrade();
        });
        binding.toolbar.setNavigationOnClickListener(_ -> supportFinishAfterTransition());
        fragment = binding.fragment.getFragment();
        fragment.setScrollBottom(() -> {
            if ((page - 1) * 10 < total)
                getGrade();
        });
        binding.toolbar.getMenu().add("导出").setIcon(R.drawable.export).setOnMenuItemClickListener(_ -> {
            startActivity(new Intent(this, MarkdownViewActivity.class).putExtra("content", fragment.toTable()).putExtra("title", "成绩"),
                    ActivityOptionsCompat.makeSceneTransitionAnimation(this, binding.toolbar, "miniapp").toBundle());
            return false;
        }).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
        yearPop = new PopupMenu(this, binding.year);
        trainTypePop = new PopupMenu(this, binding.trainType);
        courseTypePop = new PopupMenu(this, binding.courseType);
        PreferenceEditBinding courseNameEditText = PreferenceEditBinding.inflate(getLayoutInflater());
        PopupWindow courseNamePop = getPopupWindow(courseNameEditText);
        binding.year.setOnClickListener(_ -> yearPop.show());
        binding.trainType.setOnClickListener(_ -> trainTypePop.show());
        binding.courseType.setOnClickListener(_ -> courseTypePop.show());
        binding.courseName.setOnClickListener(v -> {
            input = courseName;
            courseNamePop.showAsDropDown(v);
            courseNameEditText.textInputLayout.setHint(R.string.course_name);
            courseNameEditText.textInputLayout.requestFocus();
            courseNameEditText.textField.setText(courseName.getValue());
        });
        binding.courseNumber.setOnClickListener(v -> {
            input = courseNumber;
            courseNamePop.showAsDropDown(v);
            courseNameEditText.textInputLayout.requestFocus();
            courseNameEditText.textInputLayout.setHint(R.string.course_number);
            courseNameEditText.textField.setText(courseNumber.getValue());
        });
        binding.minGrade.setOnClickListener(v -> {
            input = minGrade;
            courseNamePop.showAsDropDown(v);
            courseNameEditText.textInputLayout.requestFocus();
            courseNameEditText.textInputLayout.setHint(R.string.min_grade);
            courseNameEditText.textField.setText(minGrade.getValue());
        });
        year.observe(this, _ -> regetGrade());
        trainType.observe(this, _ -> regetGrade());
        courseType.observe(this, _ -> regetGrade());
        courseName.observe(this, s -> {
            binding.courseName.setText(s.isEmpty() ? getString(R.string.course_name) : s);
            regetGrade();
        });
        courseNumber.observe(this, s -> {
            binding.courseNumber.setText(s.isEmpty() ? getString(R.string.course_number) : s);
            regetGrade();
        });
        minGrade.observe(this, s -> {
            binding.minGrade.setText(s.isEmpty() ? getString(R.string.min_grade) : s);
            regetGrade();
        });
        getGrade();
        IntStream.range(0, 3).forEach(this::getData);
        model.getMessage().observe(this, message -> {
            JSONObject response = message.second;
            if (response.getInteger("code") == 200) {
                int what = message.first;
                switch (what) {
                    case 3 -> {
                        if (total == -1)
                            total = response.getJSONObject("data").getInteger("total");
                        response.getJSONObject("data").getJSONArray("rows").forEach(item -> fragment.add(((JSONObject) item).getString("courseName"), List.of("绩点", "教学班编号", "课程类别", "课程ID", "课程名称", "课程编号", "学分", "考试性质", "等级", "年级", "开设单位", "学期", "总学时", "培养类别", "总成绩"),
                                extractValue((JSONObject) item, new String[]{"achievementPoint", "classesNum", "courseCategoryName", "courseId", "courseName", "courseNum", "credit", "examNatureName", "finalAchievementStr", "grade", "openClassUnitName", "schoolSemester", "sumHours", "trainingCategoryName", "totalAchievement"})));
                    }
                    case 0, 1, 2 -> {
                        Menu menu = List.of(trainTypePop, yearPop, courseTypePop).get(what).getMenu();
//                        menu.dispose();
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P)
                            menu.setGroupDividerEnabled(true);
                        MutableLiveData<String> realLiveDataValue = List.of(trainType, year, courseType).get(what);
                        MaterialButton button = List.of(binding.trainType, binding.year, binding.courseType).get(what);
                        menu.add(0, 0, 0, R.string.reset).setOnMenuItemClickListener(_ -> {
                            button.setText(List.of(R.string.train_type, R.string.year, R.string.course_type).get(what));
                            realLiveDataValue.setValue("");
                            return true;
                        });
                        response.getJSONArray("data").forEach(e -> menu.add(1, 0, 0, ((JSONObject) e).getString(new String[]{"dataName", "acadYearSemester", "catName"}[what])).setOnMenuItemClickListener(item -> {
                            String menuValue = ((JSONObject) e).getString(new String[]{"dataNumber", "acadYearSemester", "catCode"}[what]);
                            if (!Objects.equals(menuValue, realLiveDataValue.getValue())) {
                                button.setText(item.getTitle());
                                realLiveDataValue.setValue(menuValue);
                            }
                            return true;
                        }));
                    }
                }
                model.nextAll();
            }
        });
    }
    
    private PopupWindow getPopupWindow(PreferenceEditBinding courseNameEditText) {
        PopupWindow courseNamePop = new PopupWindow(this, null, rikka.preference.simplemenu.R.attr.popupMenuStyle);
        courseNamePop.setContentView(courseNameEditText.getRoot());
        courseNamePop.setFocusable(true);
        courseNamePop.setOutsideTouchable(false);
        courseNamePop.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        courseNamePop.setInputMethodMode(PopupWindow.INPUT_METHOD_NEEDED);
        courseNamePop.setWidth(-1);
        courseNamePop.setOnDismissListener(() -> {
            String text = toStringOrDefault(courseNameEditText.textField.getText());
            if (!Objects.equals(input.getValue(), text))
                input.setValue(text);
        });
        return courseNamePop;
    }
    
    private void regetGrade() {
        clear();
        getGrade();
    }
    
    private void clear() {
        fragment.clear();
        page = 1;
        total = -1;
    }
    
    void getGrade() {
        model.addAndNext("jwxt/achievement-manage/achievement/selfPageList", String.format("{\"pageNo\":%s,\"pageSize\":10,\"total\":true,\"param\":%s}", page++, getArgs().toString()), 3);
    }

//    void getYear() {
//        model.addAndNext("jwxt/base-info/acadyearterm/findAcadyeartermNamesBox", 0);
//    }
//
//    void getTrainingType() {
//        model.addAndNext("jwxt/base-info/codedata/findcodedataNames?datableNumber=97", 1);
//    }
//
//    void getCourseType() {
//        model.addAndNext("jwxt/base-info/base-category/SfqyBox", 2);
//    }
    
    void getData(int pos) {
        model.add(List.of("jwxt/base-info/codedata/findcodedataNames?datableNumber=97",
                "jwxt/base-info/acadyearterm/findAcadyeartermNamesBox",
                "jwxt/base-info/base-category/SfqyBox").get(pos), pos);
    }
    
    /*
     * {"categoryCode":"01","schoolSemester":"2025-1","courseTypeCode":"10","courseNum":"编码","courseName":"名称","finalAchievement":0,"achievementState":null}
     * */
    JSONObject getArgs() {
        JSONObject args = new JSONObject();
        if (!isEmpty(trainType.getValue()))
            args.put("categoryCode", trainType.getValue());
        if (!isEmpty(year.getValue()))
            args.put("schoolSemester", year.getValue());
        if (!isEmpty(courseType.getValue()))
            args.put("courseTypeCode", courseType.getValue());
        if (!isEmpty(courseName.getValue()))
            args.put("courseName", courseName.getValue());
        if (!isEmpty(courseNumber.getValue()))
            args.put("courseNum", courseNumber.getValue());
        if (!isEmpty(minGrade.getValue()))
            args.put("finalAchievement", Integer.parseInt(minGrade.getValue()));
        return args;
    }
}

File: app/src/main/java/com/sysu/edu/academic/GradeForLevelActivity.kt
package com.sysu.edu.academic

import android.content.Intent
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.lifecycle.ViewModelProvider
import com.sysu.edu.BaseActivity
import com.sysu.edu.R
import com.sysu.edu.api.CommonUtil.extractValue
import com.sysu.edu.api.DataStoreManager
import com.sysu.edu.browser.RichTextActivity
import com.sysu.edu.view.ActivityPager
import com.sysu.edu.view.InputDialogChip
import com.sysu.edu.view.SectionData
import com.sysu.edu.view.SingleSelectChipDropdown
import com.sysu.edu.view.StaggerScreen
import com.sysu.edu.view.toMarkdown

class GradeForLevelActivity : BaseActivity() {
	@OptIn(ExperimentalLayoutApi::class) override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		enableEdgeToEdge()
		val viewModel = ViewModelProvider(this)[GradeForLevelViewModel::class.java]
		
		setContent {
			val gradeList by viewModel.gradeList.observeAsState(emptyList())
			val trainTypeOptions by viewModel.trainTypeOptions.observeAsState(emptyList())
			val yearOptions by viewModel.yearOptions.observeAsState(emptyList())
			val courseTypeOptions by viewModel.courseTypeOptions.observeAsState(emptyList())
			
			LaunchedEffect(Unit) {
				viewModel.fetchOptions()
				viewModel.fetchGrade()
			}
			var trainTypeValue by remember { mutableStateOf<String?>(null) }
			var yearValue by remember { mutableStateOf<String?>(null) }
			var courseTypeValue by remember { mutableStateOf<String?>(null) }
			var courseNameValue by remember { mutableStateOf("") }
			var courseNumberValue by remember { mutableStateOf("") }
			var minGradeValue by remember { mutableStateOf("") }
			fun onFilterChange() {
				viewModel.trainType = trainTypeValue
				viewModel.year = yearValue
				viewModel.courseType = courseTypeValue
				viewModel.courseName = courseNameValue.ifEmpty { null }
				viewModel.courseNumber = courseNumberValue.ifEmpty { null }
				viewModel.minGrade = minGradeValue.ifEmpty { null }
				viewModel.reFetchGrade()
			}
			
			val sections = remember(gradeList) {
				mutableStateListOf<SectionData>().also { list ->
					gradeList.forEach { item ->
						list.add(SectionData(title = item.getString("courseName"),
						                     rows = extractValue(this@GradeForLevelActivity,
						                                         item,
						                                         intArrayOf(R.string.gpa,
						                                                    R.string.class_number,
						                                                    R.string.course_category,
						                                                    R.string.course_id,
						                                                    R.string.course_name,
						                                                    R.string.course_number,
						                                                    R.string.credit,
						                                                    R.string.exam_nature,
						                                                    R.string.level,
						                                                    R.string.grade,
						                                                    R.string.department,
						                                                    R.string.semester,
						                                                    R.string.total_hours,
						                                                    R.string.training_category,
						                                                    R.string.total_achievement),
						                                         arrayOf("achievementPoint",
						                                                 "classesNum",
						                                                 "courseCategoryName",
						                                                 "courseId",
						                                                 "courseName",
						                                                 "courseNum",
						                                                 "credit",
						                                                 "examNatureName",
						                                                 "finalAchievementStr",
						                                                 "grade",
						                                                 "openClassUnitName",
						                                                 "schoolSemester",
						                                                 "sumHours",
						                                                 "trainingCategoryName",
						                                                 "totalAchievement"))))
					}
				}
			}
			
			ActivityPager(title = stringResource(R.string.grade_for_level), onNavigationClick = { supportFinishAfterTransition() }, isNestedScrollEnabled = false, actions = {
				IconButton(onClick = {
					val markdown = sections.toMarkdown()
					DataStoreManager.saveContent(this@GradeForLevelActivity, getString(R.string.grade_for_level), markdown) {
						startActivity(Intent(this@GradeForLevelActivity, RichTextActivity::class.java).putExtra("type", DataStoreManager.ContentType.MARKDOWN.name).putExtra("title", getString(R.string.grade_for_level)))
					}
				}) {
					Icon(painter = painterResource(R.drawable.export), contentDescription = stringResource(R.string.export))
				}
			}) {
				Column(modifier = Modifier.fillMaxSize()) {
					FlowRow(modifier = Modifier
						.fillMaxWidth()
						.padding(dimensionResource(R.dimen.horizontal_padding), dimensionResource(R.dimen.vertical_padding)),
					        horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.horizontal_margin))) {
						SingleSelectChipDropdown(category = stringResource(R.string.train_type),
						                         options = listOf(getString(R.string.reset)) + trainTypeOptions.map { it.getString("dataName") },
						                         optionValues = listOf(null) + trainTypeOptions.map { it.getString("dataNumber") },
						                         selectedValue = trainTypeValue,
						                         onValueChange = { trainTypeValue = it; onFilterChange() })
						SingleSelectChipDropdown(category = stringResource(R.string.year),
						                         options = listOf(getString(R.string.reset)) + yearOptions.map { it.getString("acadYearSemester") },
						                         optionValues = listOf(null) + yearOptions.map { it.getString("acadYearSemester") },
						                         selectedValue = yearValue,
						                         onValueChange = { yearValue = it; onFilterChange() })
						SingleSelectChipDropdown(category = stringResource(R.string.course_type),
						                         options = listOf(getString(R.string.reset)) + courseTypeOptions.map { it.getString("catName") },
						                         optionValues = listOf(null) + courseTypeOptions.map { it.getString("catCode") },
						                         selectedValue = courseTypeValue,
						                         onValueChange = { courseTypeValue = it; onFilterChange() })
						InputDialogChip(stringResource(R.string.course_name), courseNameValue) { courseNameValue = it; onFilterChange() }
						InputDialogChip(stringResource(R.string.course_number), courseNumberValue) { courseNumberValue = it; onFilterChange() }
						InputDialogChip(stringResource(R.string.min_grade), minGradeValue, KeyboardType.Number) { minGradeValue = it; onFilterChange() }
					}
					
					StaggerScreen(sections = sections, onScrollBottom = {
						if (viewModel.hasMore()) viewModel.fetchGrade()
					})
				}
			}
		}
	}
}